proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,150 @@
+Penalizing this behaviour with runtime fees collected by validators can create a perverse 
+incentive to artificially delay HFT transactions, cause them to fail and charge penalty fees.
+
+## Solution
+
+As a high-performance cluster, we want to incentivize players which feed the network 
+with correctly formed transactions instead of spamming the network. Introduction 
+of application fees would be interesting way to penalize the bad actors and applications 
+can rebate these fees to the good actors. This means the application has to decide who to 
+rebate and who to penalize through lamport transfers. In particular it needs to be able to 
+penalize, even if the application is not CPI'd to. There are multiple possible approaches 
+to provide the application with access to transfer lamports outside of the regular cpi 
+execution context:
+
+1. A new specialized fee collection mechanism that uses per account meta-data to encode 
+additional fees. Lamports should be collected on the actual accounts until claimed by 
+their owner. Account owners can trigger a per account rebate on the fee collected 
+during a transaction. Two strategies have been proposed:
+
+    1. Create PDAs of the application program to store fee settings
+
+        1. Concern is well encapsulated and implementation can be easily modified (**+**)
+        1. Overhead for loading a lot of PDAs might be high (**-**)
+
+    1. Extend existing account structure in ledger to store fee settings and collect lamports
+
+        1. High efficiency, minimal performance impact (**+**)
+        1. Accounts structure is difficult to modify (**-**)
+    
+    1. Extending accounts db to create a new type of accounts called "Cached accounts",
+These accounts will be always cached by the solana validator. These accounts can have constaints
+like small size and much more rent required to create.
+
+2. A generic execution phase that gets invoked for every program account passed to a 
+transaction. Programs would optionally contain a fee entry-point in their binary 
+code that gets invoked with a list of all accessed account keys and their lock 
+status. Programs would need access to a new sysvar identifying the transaction fee 
+payer to rebate lamports to, to prevent breaking API changes for clients.
+
+    1. Highest degree of flexibility for programs (**+**)
+    1. Least data structures modified in runtime (**+**)
+    1. Does not allow guarding non-pda accounts (e.g. an end-users signer key) (**-**)
+    1. Allowing program execution that continues on failure might invite users to 
+implement unforseen side-effects inside the fee entry-point (**-**)
+
+
+### Application fees in working
+
+As an owner of a writable account that is used a lot in the network, a program can use 
+the application fees program to assign it an application fee. This application fee is 
+applied to every transaction which will try to lock the account in writable mode. 
```

Comment:
Taking a read-lock on the account would block other users from writing. What was the thought process for making these write-only fees?

Reply:
There is currently not enough economic incentive to do that, hence it's not an issue in production, hence no one is trying to fix it.

Reply:
New Proposal will work with read locks too.